CS483/CS483/Kartaclysm/Components/ComponentTrack.cpp

```diff
@@ -0,0 +1,234 @@
+
+			HeatStroke::Event* pEvent = new HeatStroke::Event("RacerTrackPieceUpdated");
+			pEvent->SetStringParameter("racerId", m_vRacers[iRacerIndex].id);
+			pEvent->SetStringParameter("TrackPieceId", m_vTrackPieces[iNextTrackIndex]->GetGUID());
+			HeatStroke::EventManager::Instance()->TriggerEvent(pEvent);
+		}
+
+		UpdateRacerPositions();
+
+		PrintRacerPositions();
+	}
+
+	void ComponentTrack::OnRacerTrackPieceCollision(const HeatStroke::Event* p_pEvent)
+	{
+		// parse event data
+		std::string strRacerId = "";
+		std::string strTrackPieceId = "";
+		p_pEvent->GetRequiredStringParameter("racerId", strRacerId);
+		p_pEvent->GetRequiredStringParameter("TrackPieceId", strTrackPieceId);
+
+		// get racer and track piece
+		int iRacerIndex = GetRacerIndex(strRacerId);
+		int iTrackPieceIndex = GetTrackPieceIndex(strTrackPieceId);
+		if (iTrackPieceIndex < 0 || iTrackPieceIndex >= m_vTrackPieces.size() || iRacerIndex < 0)
+		{
+			return;
+		}
+
+		// update lap/track piece information
+		if (iTrackPieceIndex == 0 && m_vRacers[iRacerIndex].currentTrackPiece == m_vTrackPieces.size() - 1)
+		{
+			m_vRacers[iRacerIndex].currentTrackPiece = 0;
+			m_vRacers[iRacerIndex].currentLap++;
+		}
+		else if (iTrackPieceIndex == m_vRacers[iRacerIndex].currentTrackPiece + 1)
+		{
+			m_vRacers[iRacerIndex].currentTrackPiece = iTrackPieceIndex;
+		}
+	}
+
+	int ComponentTrack::GetTrackPieceIndex(const std::string& p_strTrackPieceId)
+	{
+		const char* strTrackPieceId = p_strTrackPieceId.c_str();
+		for (unsigned int i = 0; i < m_vTrackPieces.size(); ++i)
+		{
+			if (strcmp(strTrackPieceId, m_vTrackPieces[i]->GetGUID().c_str()) == 0)
+			{
+				return i;
+			}
+		}
+
+		return -1;
+	}
+
+	int ComponentTrack::GetNextTrackPieceIndex(int p_iCurrentTrackPieceIndex)
+	{
+		int iNextTrackPieceIndex = p_iCurrentTrackPieceIndex + 1;
+		return iNextTrackPieceIndex >= m_vTrackPieces.size() ? 0 : iNextTrackPieceIndex;
+	}
+
+	int ComponentTrack::GetRacerIndex(const std::string& p_strRacerId)
+	{
+		const char* strRacerId = p_strRacerId.c_str();
+		for (unsigned int i = 0; i < m_vRacers.size(); ++i)
+		{
+			if (strcmp(strRacerId, m_vRacers[i].id.c_str()) == 0)
+			{
+				return i;
+			}
+		}
+
+		return -1;
+	}
+
+	void ComponentTrack::UpdateRacerPositions()
+	{
+		// TODO: optimize
+		// Matt: This is currently O(n^2).  While n will always be small, and it likely won't be a problem,
+		//			this can still be improved.  Something to pick up on a rainy day.
+		for (int i = 1; i < m_vRacers.size(); ++i)
+		{
+			RacerData racerA = m_vRacers[i];
+			bool bRacerPositionUpdated = false;
+			for (int j = i - 1; j >= 0; --j)
+			{
+				RacerData racerB = m_vRacers[j];
+				if (IsAhead(racerA, racerB))
+				{
+					m_vRacers[i] = racerB;
+					m_vRacers[j] = racerA;
+					bRacerPositionUpdated = true;
+					TriggerRacerPositionUpdateEvent(racerB.id);
+				}
+			}
+
+			if (bRacerPositionUpdated)
+			{
+				TriggerRacerPositionUpdateEvent(racerA.id);
+			}
+		}
+	}
+
```

Comment:
I can add a task for this sprint to compare the performance of different algorithms with various numbers of racers so we can figure out what we want to go with. Will stick with this for now, as it's easy and doesn't have any immediately visible performance costs